packages/privacy-grade/src/classes/trackers.js

```diff
@@ -0,0 +1,593 @@
+        })
+        return surrogateList
+    }
+
+    /**
+     * @param {string} url
+     * @returns {{fromCname: string | undefined, finalURL: string}}
+     */
+    resolveCname (url) {
+        const parsed = this.tldts.parse(url, { allowPrivateDomains: true })
+        let finalURL = url
+        let fromCname
+        if (parsed && this.cnames && parsed.domain) {
+            let domain = parsed.domain
+            if (parsed.subdomain) {
+                domain = parsed.subdomain + '.' + domain
+            }
+            const finalDomain = this.cnames[domain] || domain
+            finalURL = finalURL.replace(domain, finalDomain)
+            if (finalDomain !== domain) {
+                fromCname = domain
+            }
+        }
+        return {
+            fromCname,
+            finalURL
+        }
+    }
+
+    /**
+     * Copied from extension (FIX)
+     * @param {string} urlString
+     **/
+    getBaseDomain (urlString) {
+        const parsedUrl = this.tldts.parse(urlString, { allowPrivateDomains: true })
+        return parsedUrl.domain || parsedUrl.hostname
+    }
+
+    /**
+     * single object with all of our request and site data split and
+     * processed into the correct format for the tracker set/get functions.
+     * This avoids repeat calls to split and util functions.
+     * @param {string} urlToCheck
+     * @param {string} siteUrl
+     * @param {RequestExpression} request
+     * @returns {RequestData | null}
+     */
+    getRequestData (urlToCheck, siteUrl, request) {
+        const siteDomain = this.getBaseDomain(siteUrl)
+        const urlToCheckDomain = this.getBaseDomain(urlToCheck)
+        if (!siteDomain || !urlToCheckDomain) {
+            return null
+        }
+        return {
+            siteUrl: siteUrl,
+            request: request,
+            sameBaseDomain: siteDomain === urlToCheckDomain,
+            siteDomain,
+            siteUrlSplit: this.utils.extractHostFromURL(siteUrl).split('.'),
+            urlToCheck: urlToCheck,
+            urlToCheckDomain,
+            urlToCheckSplit: this.utils.extractHostFromURL(urlToCheck).split('.')
+        }
+    }
+
+    /**
+     * @param {string} url
+     * @returns {boolean}
+     */
+    isSpecialURL (url) {
+        let urlObj
+        try {
+            urlObj = new URL(url)
+        } catch {
+            // This really shouldn't happen but if it does, we'll just assume it's a special URL
+            return true
+        }
+        const specialProtocols = [
+            // Browser specific internal protocols
+            'chrome-extension:',
+            'chrome-devtools:',
+            'chrome-search:',
+            'chrome:',
+            'edge:',
+            'opera:',
+            'about:',
+            'moz-extension:',
+
+            // Special web protocols
+            'file:',
+            'javascript:',
+            'data:',
+            'blob:',
+            'view-source:',
+            'vbscript:',
```

Comment:
Is this even supported in any browsers?

Reply:
This was triggered by our code scanning. No idea what the attack vector would be here.
![Screenshot 2023-03-08 at 13 41 30](https://user-images.githubusercontent.com/248111/223715977-ee067a8f-4796-4d2a-8a3f-bf765d638db1.png)

Reply:
Ah, I mean we can leave it in but yeah this iiuc isn't relevant to our use-case at all.